File: app/lib/assets_loader.py
"""
Carrega os assets de marca das Lojas MSC (mascote, logos).
Procura na pasta msc_brand (pasta pai do app).
Faz remocao automatica de fundo branco quando o PNG nao tem alpha.
"""
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont, ImageFilter
import os

ROOT = Path(__file__).parent.parent
BRAND_DIR = ROOT.parent

ASSET_ALIASES = {
    "mascote":    ["mascote", "mascot", "vovo", "papai_msc"],
    "logo_dark":  ["logo_msc", "logo_azul", "logo_dark", "logo"],
    "logo_white": ["logo_white", "logo_branco", "logo_branca", "msc_branco"],
}

LATO_BLACK = None
LATO_BOLD  = None
for p in [
    "C:/Windows/Fonts/Lato-Black.ttf",
    "C:/Windows/Fonts/arialbd.ttf",
    "C:/Windows/Fonts/segoeuib.ttf",
    "/usr/share/fonts/truetype/lato/Lato-Black.ttf",
]:
    if os.path.exists(p):
        LATO_BLACK = p
        break
for p in [
    "C:/Windows/Fonts/Lato-Bold.ttf",
    "C:/Windows/Fonts/arialbd.ttf",
    "/usr/share/fonts/truetype/lato/Lato-Bold.ttf",
]:
    if os.path.exists(p):
        LATO_BOLD = p
        break

if not LATO_BLACK:
    LATO_BLACK = LATO_BOLD or "arial.ttf"
if not LATO_BOLD:
    LATO_BOLD = LATO_BLACK


def _scan_dir(d):
    if not d.exists():
        return {}
    found = {}
    for p in d.iterdir():
        if p.is_file() and p.suffix.lower() in (".png", ".jpg", ".jpeg", ".webp"):
            key = p.stem.lower().replace(" ", "_").replace("-", "_")
            found[key] = p
    return found


def find_asset(kind):
    aliases = ASSET_ALIASES.get(kind, [kind])
    excludes = ["white", "branco", "branca"] if kind == "logo_dark" else []
    candidates = []
    for d in [BRAND_DIR]:
        for name, path in _scan_dir(d).items():
            if any(x in name for x in excludes):
                continue
            for alias in aliases:
                if alias in name:
                    candidates.append(path)
                    break
    if not candidates:
        return None
    pngs = [p for p in candidates if p.suffix.lower() == ".png"]
    return pngs[0] if pngs else candidates[0]


def _has_alpha(img):
    if img.mode != "RGBA":
        return False
    return img.split()[-1].getextrema()[0] < 250


def _trim(img):
    bbox = img.getbbox()
    return img.crop(bbox) if bbox else img


def _remove_white_bg(img, threshold=240, feather=2):
    img = img.convert("RGBA")
    px = img.load()
    w, h = img.size
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            if r >= threshold and g >= threshold and b >= threshold:
                m = (r + g + b) / 3
                t = max(0, min(1, (m - threshold) / max(1, 255 - threshold)))
                px[x, y] = (r, g, b, int(a * (1 - t)))
    if feather:
        alpha = img.split()[-1].filter(ImageFilter.GaussianBlur(feather))
        img.putalpha(alpha)
    return _trim(img)


def _smart_load(path, threshold=240, feather=1):
    img = Image.open(path).convert("RGBA")
    if _has_alpha(img):
        return _trim(img)
    return _remove_white_bg(img, threshold, feather)


def _resize_h(img, target_h):
    if not target_h:
        return img
    r = target_h / img.size[1]
    return img.resize((max(1, int(img.size[0] * r)), target_h), Image.LANCZOS)


def load_mascote(target_h=None):
    p = find_asset("mascote")
    if p:
        return _resize_h(_smart_load(p, threshold=248, feather=1), target_h)
    return Image.new("RGBA", (10, 10), (0, 0, 0, 0))


def load_logo(variant="dark", target_h=None):
    key = "logo_white" if variant == "white" else "logo_dark"
    p = find_asset(key)
    if p:
        return _resize_h(_smart_load(p, threshold=235, feather=1), target_h)
    return Image.new("RGBA", (10, 10), (0, 0, 0, 0))


def asset_status():
    return {
        "mascote":    find_asset("mascote"),
        "logo_dark":  find_asset("logo_dark"),
        "logo_white": find_asset("logo_white"),
    }


if __name__ == "__main__":
    for k, v in asset_status().items():
        print(f"  {k:12} -> {v if v else 'AUSENTE'}")

File: studio/src/lib/renderer/assets-loader.ts
/**
 * Carrega assets de marca (mascote, logo) com remocao automatica de fundo branco.
 * Procura nas pastas: ASSETS_DIR (env), pasta pai do projeto, e public/brand/.
 */
import path from "path";
import fs from "fs";
import sharp from "sharp";
import { Image as CanvasImage } from "@napi-rs/canvas";
import { removeWhiteBg, trimAlpha } from "./canvas-utils";

const ASSET_ALIASES: Record<string, string[]> = {
  mascote: ["mascote", "mascot", "vovo", "papai_msc"],
  logo_dark: ["logo_msc", "logo_azul", "logo_dark", "logo"],
  logo_white: ["logo_white", "logo_branco", "logo_branca", "msc_branco"],
};

function searchDirs(): string[] {
  const dirs: string[] = [];
  if (process.env.ASSETS_DIR) {
    dirs.push(path.resolve(process.cwd(), process.env.ASSETS_DIR));
  }
  dirs.push(path.resolve(process.cwd(), ".."));
  dirs.push(path.resolve(process.cwd(), "../msc_brand"));
  dirs.push(path.resolve(process.cwd(), "public/brand"));
  return dirs;
}

function scanDir(d: string): Array<{ name: string; path: string }> {
  if (!fs.existsSync(d)) return [];
  const out: Array<{ name: string; path: string }> = [];
  for (const f of fs.readdirSync(d)) {
    const ext = path.extname(f).toLowerCase();
    if (![".png", ".jpg", ".jpeg", ".webp"].includes(ext)) continue;
    const stem = path.basename(f, ext).toLowerCase().replace(/[\s-]+/g, "_");
    out.push({ name: stem, path: path.join(d, f) });
  }
  return out;
}

export function findAsset(kind: keyof typeof ASSET_ALIASES): string | null {
  const aliases = ASSET_ALIASES[kind] || [kind];
  const excludes: string[] = kind === "logo_dark" ? ["white", "branco", "branca"] : [];
  const candidates: string[] = [];
  for (const dir of searchDirs()) {
    for (const f of scanDir(dir)) {
      if (excludes.some((x) => f.name.includes(x))) continue;
      if (aliases.some((a) => f.name.includes(a))) {
        candidates.push(f.path);
      }
    }
  }
  if (!candidates.length) return null;
  // prefere PNG
  const pngs = candidates.filter((p) => p.toLowerCase().endsWith(".png"));
  return pngs[0] ?? candidates[0];
}

async function hasRealAlpha(filePath: string): Promise<boolean> {
  const meta = await sharp(filePath).metadata();
  if (!meta.hasAlpha) return false;
  // tem canal alpha mas pode ser todo 255 (opaco). Verificar.
  const stats = await sharp(filePath).stats();
  const alphaCh = stats.channels[3];
  if (alphaCh && typeof alphaCh.min === "number" && alphaCh.min < 250) return true;
  return false;
}

const cache = new Map<string, Buffer>();

export async function loadCleanAsset(kind: keyof typeof ASSET_ALIASES): Promise<Buffer | null> {
  const ck = "v1:" + kind;
  if (cache.has(ck)) return cache.get(ck)!;
  const p = findAsset(kind);
  if (!p) return null;
  let buf: Buffer;
  if (await hasRealAlpha(p)) {
    buf = await sharp(p).png().toBuffer();
    buf = await trimAlpha(buf);
  } else {
    const threshold = kind === "mascote" ? 248 : 235;
    const cleaned = await removeWhiteBg(p, threshold);
    buf = await trimAlpha(cleaned);
  }
  cache.set(ck, buf);
  return buf;
}

export async function loadCleanAssetAsCanvasImage(kind: keyof typeof ASSET_ALIASES, targetH?: number) {
  const buf = await loadCleanAsset(kind);
  if (!buf) return null;
  let finalBuf = buf;
  if (targetH) {
    finalBuf = await sharp(buf).resize({ height: targetH }).png().toBuffer();
  }
  const meta = await sharp(finalBuf).metadata();
  const img = new CanvasImage();
  img.src = finalBuf;
  return { img, width: meta.width!, height: meta.height!, buffer: finalBuf };
}

export async function assetStatus() {
  return {
    mascote: findAsset("mascote"),
    logo_dark: findAsset("logo_dark"),
    logo_white: findAsset("logo_white"),
  };
}
